src/include/miopen/general_tensor_reorder_sol.hpp

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020-2022 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_GENERAL_MIOPEN_TENSOR_REORDER_SOL_HPP
+#define GUARD_GENERAL_MIOPEN_TENSOR_REORDER_SOL_HPP
+
+#include <miopen/miopen.h>
+#include <miopen/kernel_info.hpp>
+#include <miopen/op_kernel_args.hpp>
+#include <miopen/execution_context.hpp>
+#include <vector>
+
+namespace miopen {
+
+struct GeneralReorderParam
+{
+    int tile_x{0};
+    int tile_y{0};
+    int pack_x{0};
+    int pack_y{0};
+    int ediv_x{0};
+    int ediv_y{0};
+};
+
+struct GeneralReorderSolution
+{
+    GeneralReorderSolution(const ExecutionContext& ctx_,
+                           miopenDataType_t data_type_,
+                           uint32_t dim_0_,
```

Comment:
Answer required.

Reply:
dim_0_ is the dimension of tensor, I think 2^31 is safe to cover its range.

Reply:
4GiB tensors are not rare these days. IIRC tensors use 64 bits for each dimension. Can you please change this to a suitable 64-bit type (better) or MIOPEN_THROW if dim0 exceeds MAX_UINT (at the place where 64 bits are being cut to 32 bits). Thanks.

Reply:
I wanna choose the latter solution at the moment, the reason is that 32-bit magic division used in address computation and it will take a lot of time to modify it to 64-bit while simply add a uint32_t value guarder is much easier. I'll add 64-bit type reorder as a TODO.

Reply:
Can you make it so that `MakeTensorReorderAttributes` returns empty `std::unique_ptr<TensorReorderAttributesBase>` when dim0 exceeds MAX_UINT? The semantic of this return value is "tensor transform is not applicable", which is indeed so. Then the caller can use different implementation of transform (if it exists).

Reply:
> I'll add 64-bit type reorder as a TODO.

#1476 is used to track this.

Reply:
Thanks  for the advice, I'll add this feature in next PR later this week.